raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java

```diff
@@ -546,6 +546,32 @@ public void testHandleBeginQuorumEpochAfterUserInitiatedResign(
             context.listener.currentLeaderAndEpoch());
     }
 
+    @Test
+    public void testBeginQuorumHeartbeat() throws Exception {
+        int localId = 0;
+        int remoteId1 = 1;
+        int remoteId2 = 2;
+        Set<Integer> voters = Utils.mkSet(localId, remoteId1, remoteId2);
+
+        RaftClientTestContext context = new RaftClientTestContext.Builder(localId, voters).build();
+
+        context.becomeLeader();
+        assertEquals(OptionalInt.of(localId), context.currentLeader());
+
+        // begin epoch requests should be sent out every beginQuorumEpochTimeoutMs
+        context.time.sleep(context.beginQuorumEpochTimeoutMs);
+        context.client.poll();
+        context.assertSentBeginQuorumEpochRequest(context.currentEpoch(), Utils.mkSet(remoteId1, remoteId2));
+
+        context.time.sleep(context.beginQuorumEpochTimeoutMs / 2);
+        context.client.poll();
+        context.assertSentBeginQuorumEpochRequest(context.currentEpoch(), Utils.mkSet());
+
+        context.time.sleep(context.beginQuorumEpochTimeoutMs / 2);
```

Comment:
`2 * (context.beginQuorumEpochTimeoutMs / 2)` when `context.beginQuorumEpochTimeoutMs` is an `int` is not always greater than or equal to `context.beginQuorumEpochTimeoutMs`. For example, `99 / 2 = 49` and `98` (49 * 2) is not greater than or equal to `99`.

You technically want:
```java
        int partialDelay = context.beginQuorumEpochTimeoutMs / 2;
        context.time.sleep(partialDelay);
        context.client.poll();
        context.assertSentBeginQuorumEpochRequest(context.currentEpoch(), Utils.mkSet());

        context.time.sleep(context.beginQuorumEpochTimeoutMs - partialDelay);
```

Reply:
sounds good